File: hdbm/hdbm.c
#include <malloc.h>
#include <gdbm.h>

//ccinclude
#ifndef _GDBM_H
#include <gdbm.h>
#endif

typedef struct {
    GDBM_FILE zdb;
    datum zkey,zvalue;
    int zerrno;
} hdbm;

hdbm * hdbm_open(char *fn, char *cflags, int mode);
void hdbm_close(hdbm *db);
int hdbm_insert(hdbm *db, void *key, int klen, void *value, int vlen);
int hdbm_update(hdbm *db, void *key, int klen, void *value, int vlen);
void * hdbm_get(hdbm *db, void *key, int klen);
int    hdbm_remove(hdbm *db,void *key, int klen);
void * hdbm_first(hdbm *db);
void * hdbm_next(hdbm *db);
int    hdbm_errno();
void   hdbm_sync(hdbm *db);
int    hdbm_reorg(hdbm *db);

//ccinclude

#define safefree(d) {if (d.dptr) {free(d.dptr); d.dptr=0;}}

hdbm * hdbm_open(char *fn, char *cflags, int mode) {
    int flags;
    if (mode==0) mode=0664;
    if (cflags[0]=='r') flags=GDBM_READER;
    else if (cflags[0]=='w') flags=GDBM_WRITER;
    else if (cflags[0]=='n') flags=GDBM_NEWDB;  //always new
    else if (cflags[0]=='c') flags=GDBM_WRCREAT;  //creat if necessary
    if (cflags[1]=='s') flags |= GDBM_SYNC;
    hdbm *hdbm1 = (hdbm *)malloc(sizeof(hdbm));
    hdbm1->zkey.dptr = hdbm1->zvalue.dptr = 0;
    hdbm1->zdb = gdbm_open(fn, 1024, flags, mode, 0);
    return hdbm1;
}

void hdbm_close(hdbm *db) {
    gdbm_close(db->zdb);
}

int hdbm_insert(hdbm *db, void *key, int klen, void *value, int vlen) {
    datum k,v;
    k.dptr=(char *)key;
    k.dsize=klen;
    v.dptr=(char *) value;
    v.dsize=vlen;
    //returns +1 instead of -1 for existing record
    return gdbm_store(db->zdb, k, v, GDBM_INSERT);
}

int hdbm_update(hdbm *db, void *key, int klen, void *value, int vlen) {
    datum k,v;
    k.dptr=(char *)key;
    k.dsize=klen;
    v.dptr=(char *) value;
    v.dsize=vlen;
    return gdbm_store(db->zdb, k, v, GDBM_REPLACE);
}

void * hdbm_get(hdbm *db, void *key, int klen) {
    datum k;
    if (key==0) {
        key = db->zkey.dptr;
        klen = db->zkey.dsize;
    }
    k.dptr=(char *)key;
    k.dsize=klen;
    safefree(db->zvalue);
    db->zvalue=gdbm_fetch (db->zdb, k);
    return db->zvalue.dptr;
}

int hdbm_remove(hdbm *db,void *key, int klen) {
    datum k;
    k.dptr=(char *)key;
    k.dsize=klen;
    return gdbm_delete (db->zdb, k);
}

void * hdbm_first(hdbm *db) {
    safefree(db->zkey);
    db->zkey = gdbm_firstkey(db->zdb);
    if (!db->zkey.dptr) return 0;
    return hdbm_get(db,0,0);
}

void * hdbm_next(hdbm *db) {
    void * d = db->zkey.dptr;
    if (!d) return 0;
    db->zkey = gdbm_nextkey(db->zdb,db->zkey);
    free(d);
    if (!db->zkey.dptr) return 0;
    return hdbm_get(db,0,0);
}

int hdbm_errno() {
    return gdbm_errno;
}

void hdbm_sync(hdbm *db) {
    gdbm_sync(db->zdb);
}

int hdbm_reorg(hdbm *db) {
    return gdbm_reorganize(db->zdb);
}

File: hdbm/Hdbm.hs
{-# LANGUAGE ForeignFunctionInterface #-}

module My.Hdbm (
    hdbmOpen
    ,hdbmClose
    ,hdbmInsert
    ,hdbmUpdate
    ,hdbmRemove
    ,hdbmGet
    ,hdbmFirst
    ,hdbmNext
    ,hdbmErrno
) where

import Control.Monad
import Foreign
import Foreign.C

foreign import ccall "hdbm.h hdbm_open" hdbm_open
    :: CString -> CString -> CInt -> IO (Ptr ())

foreign import ccall "hdbm.h hdbm_close" hdbmClose :: Ptr () -> IO ()

foreign import ccall "hdbm.h hdbm_insert" hdbm_insert
    :: Ptr () -> CString -> CInt -> CString -> CInt -> IO CInt

foreign import ccall "hdbm.h hdbm_update" hdbm_update
    :: Ptr () -> CString -> CInt -> CString -> CInt -> IO CInt

foreign import ccall "hdbm.h hdbm_remove" hdbm_remove
    :: Ptr () -> CString -> CInt -> IO CInt

foreign import ccall "hdbm.h hdbm_get" hdbm_get
    :: Ptr () -> CString -> CInt -> IO CString

foreign import ccall "hdbm.h hdbm_first" hdbm_first
    :: Ptr () -> IO CString

foreign import ccall "hdbm.h hdbm_next" hdbm_next
    :: Ptr () -> IO CString

foreign import ccall "hdbm.h hdbm_errno" hdbmErrno :: IO CInt
foreign import ccall "hdbm.h hdbm_sync" hdbmSync :: Ptr () -> IO ()
foreign import ccall "hdbm.h hdbm_reorg" hdbmReorg :: Ptr () -> IO CInt

hdbmOpen fn flags mode = do
    cfn <- newCString fn
    cflags <- newCString flags
    db <- hdbm_open cfn cflags (fromIntegral mode)
    free cflags
    free cfn
    when (db==nullPtr) (fail "open failed")
    return db

hdbmInsert db key value = do
    ckey <- newCString key
    cval <- newCString value
    rc <- hdbm_insert db ckey (fromIntegral $ length key + 1)
                         cval (fromIntegral $ length value + 1)
    free cval
    free ckey
    when (rc/=0) (fail "Insert failed")

hdbmUpdate db key value = do
    ckey <- newCString key
    cval <- newCString value
    rc <- hdbm_update db ckey (fromIntegral $ length key + 1)
                         cval (fromIntegral $ length value + 1)
    free cval
    free ckey
    when (rc/=0) (fail "Update failed")

hdbmRemove :: Ptr () -> String -> IO ()
hdbmRemove db key = do
    ckey <- newCString key
    rc <- hdbm_remove db ckey (fromIntegral $ length key + 1)
    free ckey
    when (rc/=0) (fail "Remove failed")

hdbmGet db key = do
    ckey <- newCString key
    rc <- hdbm_get db ckey (fromIntegral $ length key + 1)
    free ckey
    when (rc==nullPtr) (fail "Get failed")
    peekCString rc

hdbmFirst db = do
    rc <- hdbm_first db
    when (rc==nullPtr) (fail "First failed")
    peekCString rc

hdbmNext db = do
    rc <- hdbm_next db
    when (rc==nullPtr) (fail "Next failed")
    peekCString rc
